### PR TITLE
Implementation of nightmare sleep mechanics

### LINKED ISSUE #851: Missing support for "Nightmare" status effect
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened
- [x] I have checked the commit log to see if the issue has been resolved since my server was last updated
- [x] I have read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I have specified what branch this happens on branch: EDITME

**_Additional Information_** (Steps to reproduce/Expected behavior) **:** 
https://ffxiclopedia.fandom.com/wiki/Nightmare
https://www.bg-wiki.com/ffxi/Nightmare
https://www.bg-wiki.com/ffxi/Waking_Dreams

### PR DESCRIPTION
**Pull request by MowFord**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Resolves #851 .

player avatar using nightmare on a mob:
image

the `no_effect` mob is one that already had nightmare sleep.

image

after nightmare applied, see that there's the attp reduction as well as hp reduction (hp goes from 240 to 238), which implies the bio effect is successfully applied

sleep not being broken by helix:

image

Finally, a really hacky way to test the mobskill side of things:

image

Force a mob to use nightmare and see that the player gets sleep and bio and doesn't get woken by attacks

Onto the details:
- Nightmare abilities apply a sleep and a bio effect: evidence in the linked issue #851
- Nightmare sleep is unbroken by any DoT
  - NM/Mob version also is unbroken by _all_ damage sources: https://www.bg-wiki.com/ffxi/Promathia_Mission_3-5
- The Bio effect from Nightmare does not break any sleep effects

## Steps to test these changes

Run through the steps above. Code for waking a slept player is unchanged but confirming that nightmare sleep is woken up by Cures would be a plus

### CODE REVIEW

File: scripts/actions/abilities/pets/nightmare.lua [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    local bonus = xi.summon.getSummoningSkillOverCap(pet)
+    local resm = xi.mobskills.applyPlayerResistance(pet, -1, target, dINT, bonus, xi.element.DARK)
+    if resm < 0.5 then
+        skill:setMsg(xi.msg.basic.JA_MISS_2) -- resist message
+        return xi.effect.SLEEP_I
+    end
+
+    duration = duration * resm
+    if
+        target:hasImmunity(1) or
```
Comment (WinterSolstice8) [inline scripts/actions/abilities/pets/nightmare.lua:29]: `1`? is there no enum for this?

Comment (TeoTwawki) [reply scripts/actions/abilities/pets/nightmare.lua:29]: one exists in [battleentity.h](#L457) but becaus eit hasn't been added in scripts/enum yet there are both magic numbers as well commented out lua that probably shoudln't even exists https://github.com/search?q=repo%3ALandSandBoat%2Fserver+immunity+language%3ALua&type=code&l=Lua because those mobs pools should be doing the deed 

tl;dr we need to add a new file to the enum folder or start coming up with methods that share C++ enums to lau or lua enums to C++ so we stop having things in 2 places

Comment (WinterSolstice8) [reply scripts/actions/abilities/pets/nightmare.lua:29]: We can do a pass to fix immunity enum later

File: src/map/status_effect_container.cpp [hunk 2]
```diff
@@ -792,6 +792,16 @@ void CStatusEffectContainer::DelStatusEffectsByFlag(uint32 flag, bool silent)
     {
         if (PStatusEffect->HasEffectFlag(flag))
         {
+            // If this is an NM/Mob Nightmare sleep, it can be removed explictly by a cure
+            // see mobskills/nightmare.lua for full explanation
```
Comment (WinterSolstice8) [inline src/map/status_effect_container.cpp:796]: What is the mechanism that cure uses to bypass this check?

Comment (MowFord) [reply src/map/status_effect_container.cpp:796]: This block is only for deleting by `damage` flag. A cure would wake via a different means

**Review by WinterSolstice8 — APPROVED** [review]
